### PR TITLE
Add constraints support to `_optimization_history_plot`

### PR DESCRIPTION
**Pull request by hrntsm**
## Motivation

Currently, only the pareto front can visualize the results considering the constraints.
To make it easier to understand the results, we have added an optimized_history_plot to allow plots that consider the constraints.

## Description of the changes

Modified _optimization_history_plot.py to allow plots that reflect constraints.
Tests have been modified accordingly.

An example plot of the modified results is shown below.

```py
def objective(trial: optuna.Trial):
    x = trial.suggest_float("x", -15, 30)
    y = trial.suggest_float("y", -15, 30)
    v0 = 4 * x**2 + 4 * y**2
    trial.set_user_attr("constraint", [1000 - v0])
    return v0
```

r8MhiXGk

When calculating error bars(`plot_optimization_history(study, error_bar=True)`), only feasible trials are included in the calculation. 
Therefore, if an infeasible trial is included, output the following as warnings.
"Your study contains infeasible trials. In optimization history plot,error bars are calculated for only feasible trial values." 

If the constraints are strict, the error bar may not be displayed from about trial number 0 because the first feasible trial is not easily obtained.
The following example uses `trial.set_user_attr("constraint", [1000 - v0, x-10, y-10])`
Screenshot 2023-07-06 at 23 28 29

### CODE REVIEW

**Review by not522 — COMMENTED** [review]
Thank you for your PR! It's a desirable feature for supporting constrained optimization in Optuna.
I took a quick look. Please check my comment.

File: optuna/visualization/_optimization_history.py
Comment (hrntsm) [reply optuna/visualization/_optimization_history.py:31]: Thank you for your comment!
It is better that way. Will fix.
I'm checking other optuna implementations and see Enum using both `Enum` and `IntEnum`, are there any rules?

Comment (not522) [reply optuna/visualization/_optimization_history.py:31]: As far as I know, there are no strict rules for `Enum` or `IntEnum`. I think `Enum` is enough here.

File: optuna/visualization/_optimization_history.py
Comment (hrntsm) [inline optuna/visualization/_optimization_history.py:28]: I created an enum representing the state of the value

**Review by not522 — APPROVED** [review]
Thank you for your update! It looks good to me!
The test has room for improvement but can be followed up with other PRs. (It's OK adding tests in this PR if you have time.)
- Test creating `_OptimizationHistoryInfo` with infeasible trials
- Test warnings

File: optuna/visualization/_optimization_history.py
Comment (c-bata) [inline optuna/visualization/_optimization_history.py:76]: Can we simplify this logic like below?

```suggestion
            constraints = trial.system_attrs.get(_CONSTRAINTS_KEY)
            if constraints is None or all([x <= 0.0 x for x in constraints]):
                value_states.append(_ValueState.Feasible)
            elsel:
                value_states.append(_ValueState.Infeasible)
```

**Review by c-bata — COMMENTED** [review]
@hrntsm Thank you for your pull request. I left one suggestion.

On the whole, it looks good, but I'm considering whether we can refactor the `_aggregate` method. We could tackle it as a follow-up, but let me think the way and send the suggestion if I will have an idea.

File: optuna/visualization/_optimization_history.py
Comment (hrntsm) [reply optuna/visualization/_optimization_history.py:76]: This one is simpler and better. I will fix it.

**Comment by hrntsm** [comment]
Simplified _aggregate. I think it is better than before.

**Review by c-bata — COMMENTED** [review]
Looks good to me! I will approve after fixing the CI failures. Please wait until #4809 will be merged.

**Review by c-bata — APPROVED** [review]
LGTM!

**Pull request merged**
